Port to jbuilder

mseri opened the pull request:
This may require a spec file update, but it's not an issue as this repository is not pinned

mseri commented:
This can be merged but should not be released until https://github.com/xapi-project/xen-api/pull/3373 is merged

mseri commented:
The updated spec file is ready and has already been built and tested

lindig commented on http-svr/jbuild:
```diff
@@ -0,0 +1,86 @@
+    let ic = Unix.open_process_in
+        (cmd ^ " | grep -oEe '-ppx (\"([^\"\\]|\\.)+\"|\\w+)'")
+    in
+    let rec go ic acc =
```
It probably does not matter but this code is inefficient:

* `go` is not tail recursive
* It would be better to use `Buffer` to build large strings
* Below is a tail-recursive loop to read lines

```let read_lines io =
  let rec loop acc =
      match input_line io with
      | l -> loop (l :: acc)
      | exception End_of_file -> List.rev acc
  in
  loop []
```

mseri replied:
True but it is usually one or two lines at most, and it's the same boilerplate on all our jbuild files. I plan to replace it with `ppx_find` and remove it completely so I don't think it is worth touching

lindig commented on sexpr/jbuild:
```diff
@@ -0,0 +1,27 @@
+ ((name sexpr)
+  (public_name sexpr)
+  (wrapped false)
+  (modules (:standard \ sexprpp))
```
Is `\` denoting set difference?

mseri replied:
Yes

lindig commented on sexpr/sExpr.ml:
```diff
@@ -17,18 +17,30 @@ type t =
+  let open Astring in
+  String.fold_left
+    (fun acc c ->
+       acc ^
```
I am a bit worried about using string concatenation here. Performance could be a problem. I would suggest to use `Buffer`.

```
let str = Buffer.create (String.length s + 10) in
String.iter 
( function
| ... -> Buffer.add_string str "\\\\"
..
)
; Buffer.contents str
```

mseri replied:
I agree, I will modify it. Thanks

lindig commented on stunnel/stunnel.ml:
```diff
@@ -42,154 +41,155 @@ let timeoutidle = ref None
   | Nopid
 
-let string_of_pid = function
+(* let string_of_pid = function
```
Dead code?

mseri replied:
Yes, left for reference. There is a lot that could be cleaned up or simplified in the stunnel code.

lindig commented on stunnel/stunnel.ml:
```diff
@@ -42,154 +41,155 @@ let timeoutidle = ref None
+  let lines = [
+    "client=yes"; "foreground=yes"; "socket = r:TCP_NODELAY=1"; "socket = r:SO_KEEPALIVE=1"; "socket = a:SO_KEEPALIVE=1";
+    (match !timeoutidle with None -> "" | Some x -> Printf.sprintf "TIMEOUTidle = %d" x);
+    Printf.sprintf "connect=%s:%d" host port;
```
This would be better written using `List.concat` but maybe not now.

mseri replied:
Yes, but this is reindenting only for now. If somebody has time, this module could appreciate some cleanup

lindig commented on xml-light2/xml.ml:
```diff
@@ -19,189 +19,186 @@
-	done;
-	Buffer.contents buf
+  let buf = Buffer.create (String.length data + 10) in
+  for i = 0 to String.length data - 1
```
`String.iter` would be more appropriate.

mseri replied:
True, or it could be sped up as a recusive lookup that exits at the first `false`.

edwintorok commented on stunnel/stunnel.ml:
```diff
@@ -344,33 +344,29 @@ let connect
-          end
-      | [] ->
-          ()
+  let sub_after i s =
```
Would any of the substring functions from here already implement this?
http://erratique.ch/software/astring/doc/Astring.String.html#subs
http://erratique.ch/software/astring/doc/Astring.String.Sub.html

edwintorok commented on stunnel/stunnel.ml:
```diff
@@ -344,33 +344,29 @@ let connect
+      raise
+        (Stunnel_verify_error
+           (split_1 ","
+              (sub_after (e + String.length "error=") line)))
```
Not sure if this can be written differently with the Astring substring functions, but its fine as it is.

mseri replied:
It probably can, I didn't think too much about it. i think we should have an action to review and refactor stunnel, I have seen lots of things that could be simplified while I was making the jbuilder port. But again I think it should be done separately and undergo lots of testing

edwintorok commented on gzip.opam:
```diff
@@ -0,0 +1,16 @@
+opam-version: "1.2"
+maintainer: "jonathan.ludlam@eu.citrix.com"
```
are you sure we want Jon to maintain all these? or should we use the `lists.xen.org` address instead?

mseri replied:
I don't think we will publish these anyway, but I can change that to be consistent with the other libraries

edwintorok commented on xenctrlext/xenctrlext.ml:
```diff
@@ -45,7 +45,7 @@ external get_max_nr_cpus: handle -> int = "stub_xenctrlext_get_max_nr_cpus"
 
 module Xsrawext = struct
-  open Xenstore
+  (* open Xenstore *)
```
did you mean to delete the line?

mseri replied:
Yes, I did comment it to see what happens and then forgot about it. Thanks

mseri replied:
Nothing it is actually using the Xsrawext module anyway. I'll probably delete it completely and run another pair of BVT/BST

edwintorok commented on sexpr/jbuild:
```diff
@@ -14,7 +14,14 @@
  ((name sexpr)
```
Typo in commit: `s/ectract/extract`

edwintorok commented on http-svr/http.ml:
```diff
@@ -93,22 +91,25 @@ let output_http fd headers =
+let implode chr_list = String.concat "" (List.map Astring.String.of_char chr_list)
+
 let urldecode url =
-    let chars = String.explode url in
```
Should we use the `Uri` module here? Probably as a separate PR though, and after running lots of tests..

mseri replied:
I didn't know about it. Sounds like a good improvement, I'll make a ticket for it

mseri commented on gzip/gzip.ml:
```diff
@@ -12,8 +12,7 @@
+open Xapi_stdext_pervasives.Pervasiveext
 
 (** Path to the gzip binary *)
 let gzip = "/bin/gzip"
```
I don't really have the whole context to know this. It seems to me that some libraries are compatibility wrapper between old libraries and new ones, other are trying to solve problems that I was not aware of. I think we should create a new ticket to collect this information and probably write it all in a design document. What do you think about it?

edwintorok replied:
A design document sounds  good, and is independent of this PR.

mseri commented on gzip/jbuild:
```diff
@@ -0,0 +1,9 @@
+(library
+ ((name gzip)
+  (public_name gzip)
```
I didn't do it for compatibility reasons. If I prefix everything with xapi there are a number of jbuild files around that will need to be updated. I think it's better to be backward compatible for now, vendor what is vendorable from this module (most of it is used only in xapi), and update the libraries that remain.

edwintorok replied:
Sounds good, if its only used by xapi it could be deleted from here in a separate PR.

mseri commented on sexpr/sExpr.ml:
```diff
@@ -17,16 +17,16 @@ type t =
     Buffer.add_char buf '\\'
 
 let escape s =
+  let open Xapi_stdext_std.Xstringext in
```
I am all for this, but we need to figure out the corner cases first. This is used only by the xapi database, so any change will have to go under lots of testing. I'll make a ticket for it

mseri commented on sha1/sha1sum.ml:
```diff
@@ -15,9 +15,7 @@
 (** Path to the sha1sum binary (used in the new import/export code to append checksums *)
 let sha1sum = "/usr/bin/sha1sum"
```
You are probably right. I still have to check where we are using this and gzip. I'll make a separate PR to make the executables configurable

mseri replied:
I'll open a ticket for this as well

mseri commented on xml-light2/jbuild:
```diff
@@ -0,0 +1,8 @@
+(library
+ ((name xmllight2)
```
Not sure, we may need some compatibility layer. I think we should do the replacement direcly in xapi and then delete this library

mseri commented:
I have addressed the comments on this PR. Now I am going to create the ticket that we discussed but this can be merged (with its spec file update) if you are happy
